src/main/scala/sbtassembly/Assembly.scala

```diff
@@ -1,394 +1,715 @@
+import java.nio.file.{ Path, _ }
+import java.security.MessageDigest
+import java.time.Instant
+import java.util.jar.{ Attributes => JAttributes, JarFile, Manifest => JManifest }
+import scala.annotation.tailrec
+import scala.collection.GenSeq
+import scala.collection.JavaConverters._
+import scala.language.postfixOps
+import scala.tools.nsc.io.Streamable
 
+object Assembly {
   // used for contraband
-  type SeqFileToSeqFile = Seq[File] => Seq[File]
   type SeqString = Seq[String]
   type SeqShadeRules = Seq[com.eed3si9n.jarjarabrams.ShadeRule]
+  type LazyInputStream = () => InputStream
 
-  private val scalaPre213Libraries = Vector(
+  val defaultShadeRules: Seq[com.eed3si9n.jarjarabrams.ShadeRule] = Nil
+  val newLine: String = System.lineSeparator()
+  val indent: String = " " * 2
+  val newLineIndented: String = newLine + indent
+
+  private[sbtassembly] val CustomLabel = " (Custom)"
+  private[sbtassembly] type Inputs = FilesInfo[ModifiedFileInfo] :+:
+    Map[String, (Boolean, String)] :+:
+    JManifest :+:
+    Boolean :+:
+    Option[Seq[String]] :+:
+    Option[Int] :+:
+    Boolean :+:
+    HNil
+  private[sbtassembly] val scalaPre213Libraries = Vector(
     "scala-actors",
     "scala-compiler",
     "scala-continuations",
     "scala-library",
     "scala-parser-combinators",
     "scala-reflect",
     "scala-swing",
-    "scala-xml")
-  private val scala213AndLaterLibraries = Vector(
-    "scala-actors",
-    "scala-compiler",
-    "scala-continuations",
-    "scala-library",
-    "scala-reflect")
+    "scala-xml"
+  )
+  private[sbtassembly] val scala213AndLaterLibraries =
+    Vector("scala-actors", "scala-compiler", "scala-continuations", "scala-library", "scala-reflect")
 
-  val defaultExcludedFiles: Seq[File] => Seq[File] = (base: Seq[File]) => Nil
-  val defaultShadeRules: Seq[com.eed3si9n.jarjarabrams.ShadeRule] = Nil
+  /* Closeable resources */
+  private[sbtassembly] val jarFileSystemResource =
+    Using.resource((uri: URI) => FileSystems.newFileSystem(uri, Map("create" -> "true").asJava))
+  private[sbtassembly] val jarEntryInputStreamResource = Using.resource((entry: JarEntry) => entry.stream())
+  private[sbtassembly] val jarEntryOutputStreamResource = Using.resource((path: Path) =>
+    Files.newOutputStream(path, StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.CREATE)
+  )
+  private[sbtassembly] val byteArrayInputStreamResource =
+    Using.resource((bytes: Array[Byte]) => new ByteArrayInputStream(bytes))
 
-  def apply(out0: File, ao: AssemblyOption, po: Seq[PackageOption], mappings: Seq[MappingSet],
-      cacheDir: File, log: Logger): File = {
-    import Tracked.{ inputChanged, outputChanged }
-    import Cache._
-    import FileInfo.{hash, exists}
-    import java.util.jar.{Attributes, Manifest}
+  /**
+   * Represents an assembly jar entry
+   * @param target the path to write in the jar
+   * @param stream the byte payload
+   */
+  case class JarEntry(target: Path, stream: LazyInputStream) {
```

Comment:
Traditionally we use `(File, String)` pair or `Seq` of that as "mapping" - https://www.scala-sbt.org/1.x/docs/Mapping-Files.html

Using NIO `Path` might throw people off since it _sounds_ more specific than `String`, but `Path` would now be local operating system's path implementation (including Windows), and zip entry to `C:\foo\` won't make sense. So it's a mix bag because it's better typed, and worse typed at the same time.

Reply:
Hey Eugene, thanks for the comments.

Actually, I am aware of this now while building the plug-in, but `Path` actually makes sense. If you look at the `createJar` code, you can actually mount a jar/zip file as an NIO file system:

```scala
jarFileSystemResource(URI.create(s"jar:file:${output.toPath.toString}")) { jarFs =>
      val manifestPath = jarFs.getPath("META-INF/MANIFEST.MF")
```

`jarFs` is actually an NIO `FileSystem` and `jarFs.getPath` nets an NIO `Path`.

(It's a neat trick I learned and borrowed from `mill`'s assembly code.)

So  - we _can_ represent zip/jar entries as a path.

Reply:
Yea. I am aware of zipfs and NIO's ability to represent JAR entries (we borrow some of the code in Zinc to update JAR incrementally), but `Path` as a type is superset of various paths, so if a function (or in this case a case class) accepts `Path`, wouldn't it sort of imply it can handle _all_ paths? Also a `Path` representation of JAR-path would include the name of the zip or JAR file in the file system?, which too could be inconsistent with what we need, which is (relative) path entry. In other words if someone specifies `JarEntry(convertToPath(URI.create(s"jar:file:/tmp/some.jar!/foo/LICENSE")), conflicts.head.stream)`, and we're not dealing with `/tmp/some.jar`, what should sbt-assembly do? Why should we make that state representable?

Reply:
@eed3si9n 
I understand where you come from. I integrated with AWS S3's API and representing a Path is a pain. String is too generic and type-unsafe, everything can be a String. NIO Path is, as you've said above...

I am not aware of any path library that is not tied to any OS as well (such as Amazon's S3, whose root path is an empty string instead of `/`. I actually had to hand-roll my own Path implementation in that project handled via type classes to distinguish root paths of differnt file systems - but that's another story.

So this is a dilemma. 

So - what do you think we should do? Use String anyway?

Reply:
I guess my vote would be to keep it as String.

Related, sbt-assembly includes `PathList` that helps users pattern match String as delimiter-delimited Strings - https://github.com/sbt/sbt-assembly/blob/c86a0f7c1c4edd435e8c512f2716a71ed9ce8b7f/src/main/scala/sbtassembly/Assembly.scala#L387-L394

somewhat documented as https://github.com/sbt/sbt-assembly#merge-strategy